src/common/cli_wrapper.h

```diff
@@ -0,0 +1,275 @@
+#pragma once
+
+#include "3rd_party/CLI/CLI.hpp"
+#include "3rd_party/yaml-cpp/yaml.h"
+#include "common/any_type.h"
+#include "common/definitions.h"
+#include "common/logging.h"
+
+#include <iostream>
+#include <map>
+#include <string>
+
+namespace marian {
+namespace cli {
+
+// try to determine the width of the terminal
+static uint16_t guess_terminal_width(uint16_t max_width = 0,
+                                     uint16_t default_width = 180);
+
+namespace validators {
+const CLI::detail::ExistingFileValidator file_exists;
+const CLI::detail::ExistingDirectoryValidator dir_exists;
+const CLI::detail::ExistingPathValidator path_exists;
+
+const CLI::detail::NonexistentPathValidator path_not_exists;
+
+typedef CLI::Range range;
+}
+
+class CLIFormatter : public CLI::Formatter {
+public:
+  CLIFormatter(size_t columnWidth, size_t screenWidth);
+  virtual std::string make_option_desc(const CLI::Option *) const;
+
+private:
+  size_t screenWidth_{0};
+};
+
+class CLIWrapper {
+private:
+  // Stores option variables
+  std::map<std::string, Ptr<any_type>> vars_;
+  // Stores option objects
+  std::map<std::string, CLI::Option *> opts_;
+  // Command-line argument parser
+  Ptr<CLI::App> app_;
+  // Stores options as YAML object
+  YAML::Node config_;
+
+  // Name of the default option group
+  std::string defaultGroup_{""};
+  // Name of the current option group
+  std::string currentGroup_{""};
+
+  static std::string failureMessage(const CLI::App *app, const CLI::Error &e);
+
+  std::string keyName(const std::string &args) const {
+    return std::get<1>(CLI::detail::get_names(CLI::detail::split_names(args)))
+        .front();
+  }
+
+public:
+  /**
+   * @brief Create an instance of the command-line parser
+   *
+   * Option --help, -h is automatically added.
+   *
+   * @param name Header for the main option group
+   * @param columnWidth Width of the column with option names
+   * @param screenWidth Maximum allowed width for help messages
+   */
+  CLIWrapper(const std::string &name = "General options",
+             size_t columnWidth = 35,
+             size_t screenWidth = 0);
+
+  virtual ~CLIWrapper();
+
+  /**
+   * @brief Define an option with a default value
+   *
+   * @param args Comma-separated list of short and long option names
+   * @param help Help message
+   * @param val Default value
+   *
+   * @return Option object
+   */
+  template <typename T>
+  CLI::Option *add(const std::string &args, const std::string &help, T val) {
+    return add_option<T>(keyName(args), args, help, val, true, true);
+  }
+
+  /**
+   * @brief Define an option without an explicit default value. The implicit
+   * default value is T()
+   *
+   * The option will be defined in the config file even if not given as a
```

Comment:
A more general question: We have run into situations where Marian runs as a library, and models are loaded that were created before a specific parameter existed. Those will fail when the option is read out. The fix is always to add a default to the read-out operation. So in that context, should we at all have options that have implicit defaults? Why not just always use defaults in the get() calls?

Reply:
As above. We are at the moment mimicking old behavior. It's not necessarily better, just getting rid of boost.

Reply:
Would you mind adding that note to the source-code comment?

Reply:
I created an issue: #305